uid/snowid.go

```diff
@@ -0,0 +1,192 @@
+	stepBits uint8 = 12
+)
+
+const encodeBase58Map = "123456789abcdefghijkmnopqrstuvwxyzABCDEFGHJKLMNPQRSTUVWXYZ"
+
+var decodeBase58Map [256]byte
+
+// initialize mapping for decoding.
+func init() {
+	for i := 0; i < len(decodeBase58Map); i++ {
+		decodeBase58Map[i] = 0xFF
+	}
+
+	for i := 0; i < len(encodeBase58Map); i++ {
+		decodeBase58Map[encodeBase58Map[i]] = byte(i)
+	}
+}
+
+// A Node struct holds the basic information needed for a snowflake generator
+// node
+type Node struct {
+	mu    sync.Mutex
+	epoch time.Time
+	time  int64
+	node  int64
+	step  int64
+
+	nodeMax   int64
+	nodeMask  int64
+	stepMask  int64
+	timeShift uint8
+	nodeShift uint8
+}
+
+// An ID is a custom type used for a snowflake ID.  This is used so we can
+// attach methods onto the ID.
+type ID int64
+
+// NewNode returns a new snowflake node that can be used to generate snowflake
+// IDs
+func NewNode(node int64) (*Node, error) {
+	n := Node{}
+	n.node = node
+	n.nodeMax = -1 ^ (-1 << nodeBits)
+	n.nodeMask = n.nodeMax << stepBits
+	n.stepMask = -1 ^ (-1 << stepBits)
+	n.timeShift = nodeBits + stepBits
+	n.nodeShift = stepBits
+
+	if n.node < 0 || n.node > n.nodeMax {
+		return nil, errors.New("Node number must be between 0 and " + strconv.FormatInt(n.nodeMax, 10))
+	}
+
+	var curTime = time.Now()
+	// add time.Duration to curTime to make sure we use the monotonic clock if available
+	n.epoch = curTime.Add(time.Unix(epoch/1000, (epoch%1000)*1000000).Sub(curTime))
+
+	return &n, nil
+}
+
+// Generate creates and returns a unique snowflake ID
+// To help guarantee uniqueness
+// - Make sure your system is keeping accurate system time
+// - Make sure you never have multiple nodes running with the same node ID
+func (n *Node) Generate() ID {
+	n.mu.Lock()
+
+	now := time.Since(n.epoch).Nanoseconds() / 1000000
+
+	if now == n.time {
+		n.step = (n.step + 1) & n.stepMask
+
+		if n.step == 0 {
+			for now <= n.time {
+				now = time.Since(n.epoch).Nanoseconds() / 1000000
+			}
+		}
+	} else {
+		n.step = 0
+	}
+
+	n.time = now
+
+	r := ID((now)<<n.timeShift |
+		(n.node << n.nodeShift) |
+		(n.step),
+	)
+
+	n.mu.Unlock()
+	return r
+}
+
+// Bytes returns a byte array of the base58 encoded value for this ID.
+func (id ID) Bytes() []byte {
+	switch {
+	case id <= 0:
+		return nil
+	case id < 58:
+		return []byte{encodeBase58Map[id]}
+	}
+
+	b := make([]byte, 0, 11)
+	for id >= 58 {
+		b = append(b, encodeBase58Map[id%58])
+		id /= 58
+	}
+	b = append(b, encodeBase58Map[id])
+
+	for x, y := 0, len(b)-1; x < y; x, y = x+1, y-1 {
+		b[x], b[y] = b[y], b[x]
+	}
+
+	return b
+}
+
+// String returns the base58 encoded value for this ID.
+func (id ID) String() string {
+	return string(id.Bytes())
+}
+
+// Parse parses a base58 encoded value into an ID.
+func Parse(b []byte) (ID, error) {
+	switch {
+	case bytes.HasPrefix(b, []byte("1")):
+		return -1, fmt.Errorf("invalid base58: not in canonical form")
+	case len(b) > 11:
+		return -1, fmt.Errorf("invalid base58: too long")
+	}
```

Comment:
This is new. We had the length check in our wrapper around `ID`. Leading `"1"` characters are invalid because they decode into `0`, and generally leading `0` are omitted.